src/vroom_fct.cc

```diff
@@ -75,6 +75,8 @@ cpp11::integers read_fct_implicit(vroom_vec_info* info, bool include_na) {
           out[i++] = na_level;
           level_map[str.str()] = na_level;
         }
+      } else if (!include_na && matches(str, nas)) {
+        out[i++] = NA_INTEGER;
```

Comment:
I wonder if the `include_na` and `!include_na` branches could be combined. With a structure like:

```cpp
if (matches(str, nas)) {
  if (include_na) {

  } else {

  }
} else {

}
```

Reply:
I like @DavisVaughan 's suggestion. Rewriting it that way makes it easier to see that maybe you've got an unhandled case, which is leading to the malformed factor. As it stands, what happens when `include_na` is `TRUE` and you encounter an `NA` string after the first time, i.e. after you've created the `NA` level?

I think it will end up more like this:

```
    if (matches(str, nas)) {
        if (include_na) {
          if (na_level == NA_INTEGER) {
            na_level = max_level++;
            levels.push_back(NA_STRING);
            level_map[str.str()] = na_level;
          }
          out[i++] = na_level;
        } else {
          out[i++] = NA_INTEGER;
        }
      } else {
        out[i++] = max_level;
        level_map[str.str()] = max_level++;
        levels.push_back(
          info->locale->encoder_.makeSEXP(str.begin(), str.end(), false));
      }
    }
```

Reply:
I agree that Davis's edits make it look cleaner so I'm making those now and will push soon.

The loop checks a few lines up if a value already exists

```
    auto val = level_map.find(str.str());
    if (val != level_map.end()) {
      out[i++] = val->second;
    }
```
so `out` gets set further up if a level already exists (including for `NA`s).

Reply:
But there can be multiple NA strings and that is the case where the malformed factor is being created in the tests.

Reply:
Ah yes I see! Sorry, I misunderstood the context.